drivers/rtc/rtc_atmel_sam.c

```diff
@@ -0,0 +1,711 @@
+	}
+
+	if (mask & RTC_ALARM_TIME_MASK_MONTH) {
+		timeptr->tm_mon = RTC_ATMEL_SAM_REG_GET_FIELD_BCD(calalr, CALALR_MONTH) - 1;
+	}
+}
+
+static void rtc_atmel_sam_isr(const struct device *dev)
+{
+	struct rtc_atmel_sam_data *data = (struct rtc_atmel_sam_data *)dev->data;
+
+	const struct rtc_atmel_sam_config *config =
+		(const struct rtc_atmel_sam_config *)dev->config;
+
+	uint32_t sr;
+
+	sr = config->regs->RTC_SR;
+
+	if (sr & RTC_SR_ACKUPD) {
+		config->regs->RTC_SCCR = RTC_SCCR_ACKCLR;
+
+		k_sem_give(&data->cr_upd_ack_sem);
+	}
+
+	if ((sr & RTC_SR_ALARM) && (data->callback != NULL)) {
+		config->regs->RTC_SCCR = RTC_SCCR_ALRCLR;
+
+		data->callback(dev, 0, data->user_data);
+	}
+}
+
+static int rtc_atmel_sam_set_time(const struct device *dev, const struct tm *timeptr)
+{
+	struct rtc_atmel_sam_data *data = (struct rtc_atmel_sam_data *)dev->data;
+
+	const struct rtc_atmel_sam_config *config =
+		(const struct rtc_atmel_sam_config *)dev->config;
+
+	uint32_t mask_supported;
+
+	mask_supported = rtc_atmel_sam_alarm_get_supported_mask();
+
+	if (rtc_atmel_sam_validate_tm(timeptr, mask_supported) == false) {
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+
+	rtc_atmel_sam_disable_wp();
+
+	config->regs->RTC_IER = RTC_IER_ACKEN;
+
+	config->regs->RTC_CR = (RTC_CR_UPDTIM | RTC_CR_UPDCAL);
+
+	/* Await update acknowledge */
+	k_sem_take(&data->cr_upd_ack_sem, K_MSEC(1500));
+
+	config->regs->RTC_IDR = RTC_IDR_ACKDIS;
+
+	config->regs->RTC_TIMR = rtc_atmel_sam_timr_from_tm(timeptr);
+	config->regs->RTC_CALR = rtc_atmel_sam_calr_from_tm(timeptr);
+	config->regs->RTC_CR = 0;
+
+	rtc_atmel_sam_enable_wp();
+
+	k_mutex_unlock(&data->lock);
+
+	return 0;
+}
+
+static int rtc_atmel_sam_get_time(const struct device *dev, struct tm *timeptr)
+{
+	const struct rtc_atmel_sam_config *config =
+		(const struct rtc_atmel_sam_config *)dev->config;
+
+	uint32_t timr0;
+	uint32_t calr0;
+	uint32_t timr1;
+
+	/* Validate time and date */
+	if (config->regs->RTC_VER & (RTC_VER_NVTIM | RTC_VER_NVCAL)) {
+		return -ENODATA;
+	}
+
+	/* Read until synchronized (registers updated async) */
+	while (1) {
+		timr0 = config->regs->RTC_TIMR;
+		calr0 = config->regs->RTC_CALR;
+		timr1 = config->regs->RTC_TIMR;
+
+		if (timr0 == timr1) {
+			break;
+		}
+	}
+
+	timeptr->tm_sec = RTC_ATMEL_SAM_REG_GET_FIELD_BCD(timr0, TIMR_SEC);
+	timeptr->tm_min = RTC_ATMEL_SAM_REG_GET_FIELD_BCD(timr0, TIMR_MIN);
+	timeptr->tm_hour = RTC_ATMEL_SAM_REG_GET_FIELD_BCD(timr0, TIMR_HOUR);
+	timeptr->tm_mday = RTC_ATMEL_SAM_REG_GET_FIELD_BCD(calr0, CALR_DATE);
+	timeptr->tm_mon = RTC_ATMEL_SAM_REG_GET_FIELD_BCD(calr0, CALR_MONTH) - 1;
+
+	timeptr->tm_year = RTC_ATMEL_SAM_REG_GET_FIELD_BCD(calr0, CALR_YEAR);
+	timeptr->tm_year += RTC_ATMEL_SAM_REG_GET_FIELD_BCD(calr0, CALR_CENT) * 100;
+	timeptr->tm_year -= 1900;
+
+	timeptr->tm_wday = RTC_ATMEL_SAM_REG_GET_FIELD_BCD(calr0, CALR_DAY) - 1;
+	timeptr->tm_yday = -1;
+	timeptr->tm_isdst = -1;
```

Comment:
For the tm_isdst, a value less than zero indicates the data is not available https://cplusplus.com/reference/ctime/tm/, it is not explicitly stated what the tm_yday should be if not available, so i inferred it to adhere to the same rule :)

Reply:
You should probably set isdst to 0.  Otherwise anyone who uses this `tm`, for instance passes it to `gmtime()` will need to do this.  Unless they change the RTC's set value when entering/leaving DST, and no one does that.  Leaving it at -1 is just a trap for those not familiar with the POSIX time interface.

Reply:
@xyzzy42 Do you mean tm_yday? The tm_isdst must be -1 if data is not available according to the docs https://www.gnu.org/software/libc/manual/html_node/Broken_002ddown-Time.html 

I any case, RTCs can't support day light savings since the date changes depending on location, which the RTC can't know, it will always be set to -1, and is only there to be compatible with the struct tm. This is how it is in Linux as well.